### DISCUSSION TITLE
Wrong dimensions for png with orientation via xmp metadata

### DISCUSSION THREAD

**Discussion by Lumaraf** [libvips/libvips#3737] [Q&A]
**Describe the bug**
For png images with orientations >4 the Xsize and Ysize properties return the original dimensions with any metadata rotation applied. Executing `vips_autorot` does apply the correct rotation though and the orientation can also be correctly read from the metadata.
I haven't tested if this bug also occurs for pngs with orientation metadata in different formats.

**Expected behavior**
Orientation via metadata should also be applied to the dimensions for png images like it is done automatically for jpeg images.

**Environment**
Debian Bookworm running a custom build of libvips v8.14.5. All dependencies are installed from the default debian repsoitories via aptitude.
```
    meson setup build --prefix /build/libs/ --default-library static \
        -Dcplusplus=false \
        -Dmodules=disabled \
        -Dintrospection=false \
        -Dpng=enabled \
        -Djpeg=enabled \
        -Dwebp=enabled \
        -Dheif=enabled \
        -Dexif=enabled \
        -Dimagequant=enabled \
        -Dlcms=enabled
```

**Comment by jcupitt** [comment]
Hi @Lumaraf,

libvips doesn't apply orientation on load (it would kill performance). Instead, it keeps the file layout the same and carries the orientation tag through as metadata. It doesn't rotate JPEGs upright either, unless you set the `autorotate` option. `vipsthumbnail` will autorotate, but only as part of the final save operation, not as part of load.

I wonder if you are using `eog` to test the output? It has a bug whereby it will autorotate JPEGs, but won't autorotate PNGs (confusingly).

**Reply by kanekv** [reply]
@jcupitt how to get image dimensions for jpeg/png images taking into account orientation?

**Reply by jcupitt** [reply]
You can call `autorot`, then look at width/height. In python:

```python
image = pyvips.Image.new_from_file("x.png").autorot()
size = [image.width, image.height]
```

All libvips operations are "lazy" (they only compute pixels when they have to), so this is very quick.

**Comment by Lumaraf** [comment]
I don't use the autorotate option when loading jpegs and the autoration is always applied later by calling the vips_autorot method. The problem i have only applies to the Xsize and Ysize properties that are reported after loading the image and not the image contents itself.
For all the exif images from https://github.com/ianare/exif-samples/tree/master/jpg/orientation i get identical values regardless of the encoded orientation. (landscape: 1800x1200, portrait: 1200x1800) So libvips does swap the Xsize and Ysize values for orientations >4 automatically. This doesn't happen for the attached png image though which contains orientation metadata in xmp format. I get the correct orientation value via libvips though and vips_autorot also rotates the content as expected. My current problem is that i have to swap the sizes in my own code for some png images, but never for jpegs, as otherwise the calculations for the following resize and crop operations and that image would be wrong.

Note that current browsers don't support metadata orientation for png images and thus display the image wrong, this image is actually 1424x1010 if the orientation is applied.
<img width="505" alt="png-with-exif" src="https://github.com/libvips/libvips/assets/13114499/9a0471f7-97ef-4818-a85b-bc4310430894">

**Reply by jcupitt** [reply]
Ooop, sorry, I missed this post.

Your image has orientation 6 set in the PNG EXIF:

```
$ vipsheader -a orientation-6.png
orientation-6.png: 1010x1424 uchar, 4 bands, srgb, pngload
width: 1010
height: 1424
bands: 4
format: uchar
coding: none
interpretation: srgb
xoffset: 0
yoffset: 0
xres: 5.66929
yres: 5.66929
filename: orientation-6.png
vips-loader: pngload
icc-profile-data: 4064 bytes of binary data
xmp-data: 711 bytes of binary data
bits-per-sample: 8
exif-data: 150 bytes of binary data
resolution-unit: in
exif-ifd0-Orientation: 6 (Right-top, Short, 1 components, 2 bytes)
exif-ifd0-XResolution: 144/1 (144, Rational, 1 components, 8 bytes)
exif-ifd0-YResolution: 144/1 (144, Rational, 1 components, 8 bytes)
exif-ifd0-ResolutionUnit: 2 (Inch, Short, 1 components, 2 bytes)
exif-ifd2-UserComment: Screenshot (Screenshot, Undefined, 18 components, 18 bytes)
exif-ifd2-PixelXDimension: 1010 (1010, Long, 1 components, 4 bytes)
exif-ifd2-PixelYDimension: 1424 (1424, Long, 1 components, 4 bytes)
exif-ifd2-ExifVersion: Exif Version 2.1 (Exif Version 2.1, Undefined, 4 components, 4 bytes)
exif-ifd2-FlashpixVersion: FlashPix Version 1.0 (FlashPix Version 1.0, Undefined, 4 components, 4 bytes)
exif-ifd2-ColorSpace: 65535 (Uncalibrated, Short, 1 components, 2 bytes)
orientation: 6
```

You can see it's loaded as 1010 x 1424. If I apply autorot, I see:

```
$ vips autorot orientation-6.png x.png
$ vipsheader -a x.png
x.png: 1424x1010 uchar, 4 bands, srgb, pngload
width: 1424
height: 1010
bands: 4
format: uchar
coding: none
interpretation: srgb
xoffset: 0
yoffset: 0
xres: 5.66929
yres: 5.66929
filename: x.png
vips-loader: pngload
icc-profile-data: 4064 bytes of binary data
xmp-data: 711 bytes of binary data
bits-per-sample: 8
exif-data: 186 bytes of binary data
resolution-unit: in
exif-ifd0-Orientation: 1 (Top-left, Short, 1 components, 2 bytes)
exif-ifd0-XResolution: 144/1 (144, Rational, 1 components, 8 bytes)
exif-ifd0-YResolution: 144/1 (144, Rational, 1 components, 8 bytes)
exif-ifd0-ResolutionUnit: 2 (Inch, Short, 1 components, 2 bytes)
exif-ifd2-ExifVersion: Exif Version 2.1 (Exif Version 2.1, Undefined, 4 components, 4 bytes)
exif-ifd2-UserComment: Screenshot (Screenshot, Undefined, 18 components, 18 bytes)
exif-ifd2-FlashpixVersion: FlashPix Version 1.0 (FlashPix Version 1.0, Undefined, 4 components, 4 bytes)
exif-ifd2-ColorSpace: 65535 (Uncalibrated, Short, 1 components, 2 bytes)
exif-ifd2-PixelXDimension: 1424 (1424, Long, 1 components, 4 bytes)
exif-ifd2-PixelYDimension: 1010 (1010, Long, 1 components, 4 bytes)
orientation: 1
```

You can see libvips has rotated the image upright and edited the EXIF to change the orientation tag. However, the XMP data is still there, and, as you say, also contains an orientation tag:

```
$ vipsheader -f xmp-data x.png | base64 -d
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="XMP Core 5.4.0">
   <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
      <rdf:Description rdf:about=""
            xmlns:tiff="http://ns.adobe.com/tiff/1.0/"
            xmlns:exif="http://ns.adobe.com/exif/1.0/">
         <tiff:XResolution>144/1</tiff:XResolution>
         <tiff:Orientation>6</tiff:Orientation>
         <tiff:YResolution>144/1</tiff:YResolution>
         <tiff:ResolutionUnit>2</tiff:ResolutionUnit>
         <exif:PixelYDimension>1424</exif:PixelYDimension>
         <exif:UserComment>Screenshot</exif:UserComment>
         <exif:PixelXDimension>1010</exif:PixelXDimension>
      </rdf:Description>
   </rdf:RDF>
</x:xmpmeta>
```

So if you now load this into software that understands XMP, it will be rotated again.

The error is in the original image -- photoshop should not have set orientation in two places. Which should take priority? Should both be applied? It is not defined.

I think I would remove the XMP data block after autorot, or possibly edit it to also remove the orientation tag there.
